### PR TITLE
More complete Helm exporter

### PR DESCRIPTION
**Pull request by zeenix**
This PR aims to make Helm exporter as complete as possible so that it's actually useful. Using the now-added various CLI options, users can compose a commandline to export a particular habitat package and not have to hand-edit the generated chart directory, and then keep using the same commandline to create a chart for each new version and/or release.

I did `helm lint` the exported chart with various options to test a bit but I haven't tried installing the chart, which I'll do tomorrow. I wanted to create the PR already so reviews can start with I test more thoroughly.

### CODE REVIEW

File: components/pkg-export-helm/src/main.rs [hunk 1]
```diff
@@ -115,6 +115,13 @@ fn cli<'a, 'b>() -> clap::App<'a, 'b> {
                 .validator(valid_url)
                 .help("The URL of the project's home page"),
         )
+        .arg(
+            Arg::with_name("ICON")
+                .value_name("URL")
+                .long("icon")
+                .validator(valid_url)
+                .help("A URL to an SVG or PNG image to be used as an icon"),
```
Comment (asymmetric) [inline components/pkg-export-helm/src/main.rs]: I'd say "The URL of an", same as the one above.

File: components/pkg-export-helm/src/main.rs [hunk 2]
```diff
@@ -122,6 +122,9 @@ fn cli<'a, 'b>() -> clap::App<'a, 'b> {
                 .validator(valid_url)
                 .help("A URL to an SVG or PNG image to be used as an icon"),
         )
+        .arg(Arg::with_name("DEPRECATED").long("depr").help(
```
Comment (asymmetric) [inline components/pkg-export-helm/src/main.rs]: Why `depr`? I see some arguments are short, others are long. I'd argue `depr` is not catchy enough to warrant being shortened, so i'd just use `deprecate`.

File: components/pkg-export-helm/defaults/HelmChartFile.hbs [hunk 3]
```diff
@@ -1,5 +1,6 @@
 name: {{name}}
 version: {{version}}
+tillerVersion: ">=2.7.2"
```
Comment (asymmetric) [inline components/pkg-export-helm/defaults/HelmChartFile.hbs:3]: Why is this required? I would add this to the commit message and to the PR description, cause it's otherwise not clear that the PR has this side effect.

File: components/pkg-export-helm/src/main.rs [hunk 4]
```diff
@@ -108,6 +108,14 @@ fn cli<'a, 'b>() -> clap::App<'a, 'b> {
                 .long("desc")
                 .help("A single-sentence description"),
         )
+        .arg(
+            Arg::with_name("KEYWORD")
+                .value_name("KEYWORD")
+                .long("keyword")
+                .short("k")
+                .multiple(true)
+                .help("A keyword about this project"),
```
Comment (asymmetric) [inline components/pkg-export-helm/src/main.rs]: A keyword *for*, IMO.

File: components/pkg-export-helm/src/main.rs [hunk 5]
```diff
@@ -130,6 +130,15 @@ fn cli<'a, 'b>() -> clap::App<'a, 'b> {
...
                 .help("A URL to an SVG or PNG image to be used as an icon"),
         )
+        .arg(
+            Arg::with_name("SOURCE")
+                .value_name("URL")
+                .long("source")
+                .short("s")
+                .multiple(true)
+                .validator(valid_url)
+                .help("URL to source code for the project"),
```
Comment (asymmetric) [inline components/pkg-export-helm/src/main.rs]: Elsewhere you use "A URL", so you might wanna be consistent.

Also I'd say "to *the*".

File: components/pkg-export-helm/src/maintainer.rs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    email: Option<String>,
+    url: Option<String>,
+}
+
+impl Maintainer {
+    pub fn from_args(matches: &ArgMatches) -> Result<Vec<Self>> {
+        let mut maintainers = Vec::new();
+
+        if let Some(args) = matches.values_of("MAINTAINER") {
+            for arg in args {
```
Comment (asymmetric) [inline components/pkg-export-helm/src/maintainer.rs:36]: Out of curiosity, did you try doing this with `map`/`collect`?

File: components/pkg-export-helm/src/maintainer.rs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        };
+
+        Ok(maintainers)
+    }
+}
+
+impl FromStr for Maintainer {
+    type Err = Error;
+
+    fn from_str(maintainer_str: &str) -> result::Result<Self, Self::Err> {
```
Comment (asymmetric) [inline components/pkg-export-helm/src/maintainer.rs]: Maybe add a comment here showing the correct format?

File: components/pkg-export-helm/src/maintainer.rs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        let values: Vec<&str> = maintainer_str.split(',').collect();
+        if values.len() < 1 || values.len() > 3 {
+            return Err(Error::InvalidMaintainer(maintainer_str.to_owned()));
+        }
+
+        let name = values[0].to_string();
+        // FIXME: Check validity of email address
+        let email = values.get(1).map(|&s| s.to_owned());
+        let url = values.get(2).map(|&s| s.to_owned());
+        if let Some(ref u) = url {
```
Comment (asymmetric) [inline components/pkg-export-helm/src/maintainer.rs:67]: Can't you just chain this to the `map`, so that you don't need an `if`? Not sure if you decided against it for readability.

File: components/pkg-export-helm/src/maintainer.rs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            let _ = Url::parse(&u).map_err(|e| {
```
Comment (asymmetric) [inline components/pkg-export-helm/src/maintainer.rs]: Do you need the `let _`? I suppose you could just eliminate it altogether. Or does the compiler complain?

File: components/pkg-export-helm/src/maintainer.rs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            })?;
+        };
+
+        Ok(Maintainer { name, email, url })
+    }
+}
+
+impl Into<String> for Maintainer {
+    /// Convert into a string. The returned string can be directly put in a Chart.yaml file under
+    /// `maintainers` section.
```
Comment (asymmetric) [inline components/pkg-export-helm/src/maintainer.rs]: Missing *the*

File: components/pkg-export-helm/src/main.rs [hunk 11]
```diff
@@ -190,6 +203,14 @@ fn valid_url(val: String) -> result::Result<(), String> {
     }
 }
 
+fn valid_maintainer(val: String) -> result::Result<(), String> {
+    maintainer::Maintainer::from_str(&val).map(|_| ()).map_err(
```
Comment (asymmetric) [inline components/pkg-export-helm/src/main.rs]: What's the point of the first `map`?

**Review by asymmetric — CHANGES_REQUESTED** [review]

File: components/pkg-export-helm/src/main.rs (hunk 1 above)
Comment (zeenix) [reply components/pkg-export-helm/src/main.rs]: ok, about these helm strings. They are adapted copy&paste from the Helm docs: https://docs.helm.sh/developing_charts/#the-chart-yaml-file

File: components/pkg-export-helm/defaults/HelmChartFile.hbs (hunk 3 above)
Comment (zeenix) [reply components/pkg-export-helm/defaults/HelmChartFile.hbs:3]: It is a separate commit with it's own commit message but yes I should add "Require the version of tiller, helm exporter has been tested against".

File: components/pkg-export-helm/src/main.rs (hunk 11 above)
Comment (zeenix) [reply components/pkg-export-helm/src/main.rs]: there is only 1 map here and that transforms the Ok value to what the `valid_maintainer` needs to return.

File: components/pkg-export-helm/src/maintainer.rs (hunk 6 above)
Comment (zeenix) [reply components/pkg-export-helm/src/maintainer.rs:36]: Not sure but I can try..

File: components/pkg-export-helm/src/maintainer.rs (hunk 8 above)
Comment (zeenix) [reply components/pkg-export-helm/src/maintainer.rs:67]: `map`'s closure can't error out so the code in the `if` block has to be in `from_str`.

File: components/pkg-export-helm/src/maintainer.rs (hunk 9 above)
Comment (zeenix) [reply components/pkg-export-helm/src/maintainer.rs]: hmm.. actually no, it does not complain. I'll remove..

File: components/pkg-export-helm/src/maintainer.rs (hunk 6 above)
Comment (zeenix) [reply components/pkg-export-helm/src/maintainer.rs:36]: Actually same issue here. `arg.parse` can end up in error, it can't be execute as part of a `map` closure.

**Comment by zeenix** [comment]
PTALA.

**Review by asymmetric — APPROVED** [review]

**Comment by zeenix** [comment]
https://github.com/habitat-sh/habitat/pull/4715
